yew/src/virtual_dom/vtag.rs

```diff
@@ -110,6 +118,72 @@ impl Clone for VTag {
     }
 }
 
+#[cfg(feature = "ssr")]
+impl ToHtmlString for VTag {
+    fn to_html_string(&self) -> String {
+        let mut parts: Vec<String> = Vec::new();
+        let tag_name = htmlescape::encode_minimal(&self.tag).to_lowercase();
+
+        for c in tag_name.chars() {
+            let is_alnum = (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
+            if !is_alnum && c != '-' && c != '_' && c != ':' {
+                panic!("Invalid tag name: `{}`", &tag_name);
+            }
+        }
+
+        parts.push(format!("<{}", tag_name).to_string());
+
+        for (key_unclean, value) in &self.attributes {
+            let key = key_unclean.to_lowercase();
+            // checked, value (special if textarea), disabled, href?, selected,
+            // kind -> type if input, disallow ref, disallow LISTENER_SET, class
+
+            for c in key.chars() {
+                let is_alnum = (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
+                if !is_alnum && c != '-' && c != '_' && c != ':' {
+                    panic!("Invalid attribute name: `{}`", &key);
+                }
+            }
+
+            let is_skipped = tag_name == "textarea" && key == "value";
+            if !is_skipped {
+                parts.push(
+                    format!(
+                        " {}=\"{}\"",
+                        htmlescape::encode_minimal(&key),
+                        htmlescape::encode_attribute(&value)
+                    )
+                    .to_string(),
+                );
+            }
+        }
+
+        if self.checked {
+            parts.push(" checked=\"true\"".to_string())
+        }
+
+        if tag_name == "input" {
+            if let Some(kind) = &self.kind {
+                parts
+                    .push(format!(" type=\"{}\"", htmlescape::encode_attribute(&kind)).to_string());
+            }
+        }
+
+        let children_html = match tag_name.as_ref() {
+            "textarea" => VText::new(self.value.clone().unwrap_or("".to_string())).to_html_string(),
+            _ => self.children.to_html_string(),
+        };
+        if children_html == "" {
+            parts.push(" />".to_string());
+        } else {
+            parts.push(">".to_string());
+            parts.push(children_html);
+            parts.push(format!("</{}>", tag_name).to_string());
```

Comment:
The `to_string()` call shouldn't be needed here. The `format!` macro creates an expression returning a `String`. [Handy documentation on `format!`](https://doc.rust-lang.org/stable/std/macro.format.html).
```suggestion
            parts.push(format!("</{}>", tag_name));
```